cmd/cstor-pool-mgmt/volumereplica/volumereplica.go

```diff
@@ -669,3 +671,177 @@ func GetAndUpdateReplicaID(cvr *apis.CStorVolumeReplica) error {
 	}
 	return nil
 }
+
+// GetAndUpdateSnapshotInfo get the snapshot list from ZFS and updated in CVR status.
+// It doesn in following steps:
+// 1. Get snapshot list from ZFS
+// 2. Checks whether above snapshots exist on CVR under Status.Snapshots:
+//    2.1 If snapshot doesn't exist then get the info of snapshot from ZFS and update
+//        the details in CVR.Status.Snapshots
+// 3. Verify and delete the snapshot details on CVR if it is deleted from ZFS
+// 4. Update the pending list of snapshots by verifying with snapshot list obtained from step1
+// 5. If replica is under rebuilding get the snapshot list from peer CVR and update them
+//    under pending snapshot list
+func GetAndUpdateSnapshotInfo(
+	clientset clientset.Interface, cvr *apis.CStorVolumeReplica) error {
+	volName := cvr.GetLabels()[string(apis.PersistentVolumeCPK)]
+	dsName := PoolNameFromCVR(cvr) + "/" + volName
+
+	snapList, err := GetSnapshotList(dsName)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get the list of snapshots")
+	}
+
+	// Add snapshot if it doesn't exist on CVR but exist on ZFS
+	err = addSnapshotListInfo(cvr, snapList)
+	if err != nil {
+		return errors.Wrapf(err, "failed to add snapshot list info")
+	}
+
+	// Delete snapshot if exist on CVR but not in ZFS
+	for snapName, _ := range cvr.Status.Snapshots {
```

Comment:
Can we move this to separate function or merge all checks deletion/adding from snapshot,pending_snapshots to one loop?

Reply:
We can do it in a single loop.

Reply:
got to see that we are iterating over different things so can't be done in a single loop.